middle_end/flambda/inlining/inlining_decision.ml

```diff
@@ -323,71 +337,100 @@ let max_rec_depth = 1
-    in
-    let scope = DE.get_continuation_scope_level (DA.denv dacc) in
-    let _, uacc =
-      simplify_expr dacc expr ~down_to_up:(fun dacc ~rebuild ->
-        let exn_continuation = Apply.exn_continuation apply in
-        let uenv = UE.add_exn_continuation UE.empty exn_continuation scope in
-        let uenv =
-          match Apply.continuation apply with
-          | Never_returns -> uenv
-          | Return return_continuation ->
-            UE.add_return_continuation uenv return_continuation scope
-              return_arity
-          in
-          let uacc = UA.create uenv dacc in
-          rebuild uacc ~after_rebuild:(fun expr uacc -> expr, uacc)
-        )
-    in
-    UA.cost_metrics uacc
+(* XCR mshinwell: Please move speculative_inlining and might_inline to
+   toplevel, causing them to be closed by providing extra parameters, so
+   that we don't allocate closures every time we examine a call site.
+
+   poechsel: Done
+*)
+let speculative_inlining dacc ~apply ~function_decl ~simplify_expr
+      ~return_arity =
+  let dacc = DA.set_do_not_rebuild_terms_and_disable_inlining dacc in
+  (* CR poechsel: [Inlining_transforms.inline] should only be called
+     once and not twice (once there and once in [simplify_apply_expr])
+
+     mshinwell: I think it needs to be called twice; the second time,
+     inlining will be on in the environment.  Is that wrong?  Let's resolve
+     this now.
+
+     poechsel: I think you are confused between [Inlining_transforms.inline]
+     and [simplify_expr]. [Inlining_transforms] builds the expression that is
+     going to replace the call (mostly wraps the body in some continuations).
+     Right now it may be called twice for the same function (once before
+     speculating and once if we decided to inline the call in the end). If
+     needed I can make [Inlining_decision] return a
+     [Call_site_decision.t * Expr.t option] but then
+     [make_decision_for_call_site] wouldn't only be taking a decision. It may
+     remove a few allocations but I think that would be a premature optimisation.
+  *)
+  let dacc, expr =
+    (* We only speculatively inline when there's no [unroll] annotation,
+       that is the unroll_to is None *)
+    (* XCR mshinwell: I don't follow this comment.  Earlier on in this file
+       we still say that speculation can occur even if an unrolling
+       attribute was found.  Or is this comment supposed to say just that
+       we disable unrolling when speculating?
+
+       poechsel: The only way [unroll_to] is not None is when an explicit
+       Unroll annotation is provided by the user. If this is the case then
+       inliner will always inline the function and will not call
+       [speculative_inlining]. Thus inside of [speculative_inlining] we
+       will always have [unroll_to] = None. We are not disabling unrolling
+       when speculating, it just happens that no unrolling can happen while
+       speculating right now.
+    *)
```

Comment:
Please augment the comment above with your paragraph from this XCR, and delete the XCR